include/fiction/algorithms/graph/mincross.hpp

```diff
@@ -0,0 +1,624 @@
+            });
+    }
+
+    /**
+     * Reorders the nodes in a given rank according to computed medians.
+     *
+     * @param r The rank index.
+     * @param reverse If true, sorts in descending order of medians.
+     */
+    void reorder(const uint32_t r, const bool reverse)
+    {
+        // Get the nodes at rank r
+        auto rank = fanout_ntk.get_ranks(r);
+
+        // Sort by median value
+        std::sort(rank.begin(), rank.end(),
+                  [&](auto const& n1, auto const& n2)
+                  {
+                      const double m1 = median_map[n1];
+                      const double m2 = median_map[n2];
+
+                      // Handle -1 (no connections) as infinity to push to the back
+                      if (m1 == -1)
+                      {
+                          return false;
+                      }
+                      if (m2 == -1)
+                      {
+                          return true;
+                      }
+
+                      return reverse ? (m1 > m2) : (m1 < m2);
+                  });
+
+        // Re-assign sorted rank and update rank positions
+        fanout_ntk.set_ranks(r, rank);
+    }
+
+    /**
+     * Performs pairwise transpositions within ranks to further reduce crossings.
+     *
+     * @param reverse If true, applies reversed heuristic for tie-breaking.
+     */
+    void transpose(const bool reverse)
+    {
+        std::vector<bool> candidate(fanout_ntk.depth() + 1, true);
+        uint32_t          delta          = 0;
+        uint32_t          max_iterations = 0;
+
+        while (true)
+        {
+            delta = 0;
+
+            for (uint32_t l = 0; l <= fanout_ntk.depth(); ++l)
+            {
+                if (l == 0 && ps.fixed_pis)
+                {
+                    continue;
+                }
+
+                if (candidate[l])
+                {
+                    auto d = transpose_step(l, reverse);
+                    delta += d;
+
+                    candidate[l] = (d > 0);  // if changed, keep candidate
+                }
+            }
+            ++max_iterations;
+
+            if (delta < 1 || max_iterations == 1000)
+            {
+                break;
+            }
+        }
+    }
+
+    /**
+     * Performs a single transposition pass for rank `r`.
+     *
+     * @param r Rank index.
+     * @param reverse If true, applies reversed heuristic for tie-breaking.
+     * @return The number of crossings reduced.
+     */
+    uint32_t transpose_step(const uint32_t r, const bool reverse)
+    {
+        auto rank = fanout_ntk.get_ranks(r);
+
+        if (rank.size() <= 1)
+        {
+            return 0;
+        }
+
+        uint32_t rv = 0;
+
+        for (uint32_t i = 0; i < rank.size() - 1; ++i)
+        {
+            auto const& v = rank[i];
+            auto const& w = rank[i + 1];
+
+            // Ensure v comes before w
+            assert(fanout_ntk.rank_position(v) < fanout_ntk.rank_position(w));
```

Comment:
It would be a nice bit of extra information for the debugging human if we added an informative string of what exactly happened, what that means, or why this is not allowed, e.g.:
```suggestion
            assert(fanout_ntk.rank_position(v) < fanout_ntk.rank_position(w) && "v has to come before w in the rank because ...");
```